### DISCUSSION TITLE
Slow sentence encoding models

### DISCUSSION THREAD

**Discussion by ivanstepanovftw** [sonos/tract#1020] [General]
I have exported `sentence-transformers/all-MiniLM-L6-v2` (~ 60 MiB), and encoding is very slow. Encoding sentence takes 3 seconds bare metal and 10 seconds WASM, while pytorch takes about 0.03 seconds.

Bare metal results:
```
model loaded in 308 ms
model run in 3929 ms
outputs[0]: 1,128,384,F32 0.22568491, 0.23463942, 0.055242293, 0.15074176, -0.23616529, -0.10899622, 0.099527195, -0.26645622, -0.041477088, 0.023107085, 0.13349931, 0.037163362...
```

Source code:
[sentence-transformers.zip](https://github.com/sonos/tract/files/11082926/sentence-transformers.zip)

**Comment by kali** [comment]
Did you test in release ?

I get:

```
model loaded in 293 ms
model run in 66 ms
outputs[0]: 1,128,384,F32 0.22568458, 0.23463914, 0.055243142, 0.1507416, -0.2361654, -0.10899601, 0.09952721, -0.26645616, -0.04147698, 0.023107242, 0.13349918, 0.03716331...
```

**Reply by ivanstepanovftw** [reply]
Sorry, forgot to mention. Yes, it is in release mode...
```
/usr/bin/time /home/i/Documents/p/playground/tract2/target/release/sentence-transformers
model loaded in 464 ms
model run in 9021 ms
outputs[0]: 1,128,384,F32 0.22568491, 0.23463942, 0.055242293, 0.15074176, -0.23616529, -0.10899622, 0.099527195, -0.26645622, -0.041477088, 0.023107085, 0.13349931, 0.037163362...
9.23user 0.25system 0:09.53elapsed 99%CPU (0avgtext+0avgdata 281472maxresident)k
0inputs+0outputs (0major+109020minor)pagefaults 0swaps
```


For debug I get following:
```
    Finished dev [unoptimized + debuginfo] target(s) in 8.87s
     Running `/home/i/Documents/p/playground/tract/target/debug/sentence-transformers`
model loaded in 2829 ms
model run in 110270 ms
```

Strange things

**Reply by ivanstepanovftw** [reply]
For your reference, this is pytorch-albert-v2 release timings:
```
1.18user 0.53system 0:01.77elapsed 96%CPU (0avgtext+0avgdata 758720maxresident)k
0inputs+0outputs (0major+208476minor)pagefaults 0swaps
```
Rust version
```
rustc 1.68.0 (2c8cc3432 2023-03-06)
```
And pytorch results
```
from time import time
from sentence_transformers import SentenceTransformer

t1 = time()
model = SentenceTransformer('all-MiniLM-L6-v2')
dt = time() - t1
print(f"Loading model took {dt:.2f} seconds")

t1 = time()
sentence = "That is a happy person"
embedding = model.encode(sentence)
dt = time() - t1
print(f"Embedding took {dt:.2f} seconds")
```
```
Loading model took 0.31 seconds
Embedding took 0.01 seconds
```

**Reply by kali** [reply]
ok, let's see...
* you are using linux
* you are using tract main brain (up-to-date?)
* what kind of machine are you using ? x86_64 ? can you show us the output of `cat /proc/cpuinfo| grep flags | head -1` ?

Maybe you could also try a flamegraph ? https://github.com/flamegraph-rs/flamegraph ?

**Reply by ivanstepanovftw** [reply]
- Fedora Linux 37 (Workstation Edition)
- tract commit de2fb04e8aa2cef4f5f06da3ebeb7dd0cc3c9c11
- AMD Ryzen 7 5800U, plugged in, CPU Boost enabled.
    `
    flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
    `
- `cargo flamegraph --bin sentence-transformers` [flamegraph.zip](https://github.com/sonos/tract/files/11088943/flamegraph.zip) and [perf.zip](https://github.com/sonos/tract/files/11088949/perf.zip)

**Reply by kali** [reply]
ok, the flamegraph shows all this time is spent in an EinSum operators. It should have been optimised to a LirMatMulUnary. Please check:
* that you are really running the model resulting from into_optimized
* that you are on tract to of tree (there are a couple of thing that were non-deterministic that I fixed the last couple of days, which may explain why my optimised model and yours end up different.

**Reply by ivanstepanovftw** [reply]
1. Yes, I have ensured that `into_optimized()` is applied. Without into_aptimized: `model run in 186954 ms`, with: `model run in 3775 ms`
2. I am on master, latest commit was 1 day ago

**Reply by ivanstepanovftw** [reply]
Here is my work tree: https://github.com/ivanstepanovftw/tract/tree/sbert
How do I test: `cd examples/sentence_transformers`, `cargo run --release`

I can also provide you my machine using remote control apps, or SSH, or VSCode live share.

**Reply by kali** [reply]
I am not too sure why you cloned tract, but you branched from a semi-broken version, with missing optimisations. Please stick to point releases by default unless you have a good reason...

Changed `Cargo.toml` like this to use 0.19.8, and all is much better:

```
tract-onnx = "0.19.8"
```

```
model loaded in 134 ms
model run in 47 ms
```

As for the top of tree, it is still broken. Good datapoint, I was not aware some transformers were mis-behaving, so I'll try to have a look in the coming days, before cutting a 0.20. (Branch 0.19.x will not be affected)

```
tract-onnx = { git="https://github.com/sonos/tract", branch="main" }
```

gives

```
model loaded in 169 ms
model run in 3351 ms
```

Anyway, please use point releases unless you have a compelling reason :)

**Reply by ivanstepanovftw** [reply]
Thank you very much! Will try!
I am also wanted to contribute WASM example showing Sbert models usage with some linear algebra associated (mean polling and cosine similarity)

**Reply by kali** [reply]
Ha, ok, make sense to clone then, and it would be awesome to have a wasm example :) I hope I can fix the optimisation issue on main this week or the next...
